src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -336,16 +546,110 @@ private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, B
         private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, ImmutableArray<BoundExpression> loweredArgs)
         {
             Debug.Assert(loweredArgs.Length > 4);
-            Debug.Assert(loweredArgs.All(a => a.HasErrors || a.Type is { } && a.Type.IsStringType()));
+
+            var stringifiedArgs = ArrayBuilder<BoundExpression>.GetInstance(loweredArgs.Length);
+
+            foreach (var loweredArg in loweredArgs)
+            {
+                Debug.Assert(loweredArg.HasErrors || loweredArg.Type is { } argType && (argType.IsStringType() || argType.IsCharType()));
+
+                if (loweredArg.Type?.IsCharType() == true)
+                {
+                    stringifiedArgs.Add(WrapCharExprIntoToStringCall(loweredArg));
+                }
+                else
+                {
+                    stringifiedArgs.Add(loweredArg);
+                }
+            }
 
             var method = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_String__ConcatStringArray);
             Debug.Assert((object)method != null);
 
-            var array = _factory.ArrayOrEmpty(_factory.SpecialType(SpecialType.System_String), loweredArgs);
+            var array = _factory.ArrayOrEmpty(_factory.SpecialType(SpecialType.System_String), stringifiedArgs.ToImmutableAndFree());
 
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, array);
         }
 
+        private BoundExpression WrapCharExprIntoToStringCall(BoundExpression expr)
```

Comment:
>WrapCharExprIntoToStringCall [](http://example.com/codeflow?start=32&length=28)

It feels like it would be better to extract a bigger chunk of ```ConvertConcatExprToString``` into a helper that wraps any type into a ToString call. Effectively, split it in two methods right in place (which makes the diff minimal, which is very good), and without extracting ```FindSpecificToStringOfStructType```. It doesn't feel useful to have Char-specific helper, especially given that it, apparently, fails to perform the task under some conditions.